File: 3.cameraandmatrices/webgl/source/js/engine/tools/CCCameraAppUI.js
/*-----------------------------------------------------------
 * 2c - Cross Platform 3D Application Framework
 *-----------------------------------------------------------
 * Copyright © 2011 - 2011 France Telecom
 * This software is distributed under the Apache 2.0 license.
 * http://www.apache.org/licenses/LICENSE-2.0.html
 *-----------------------------------------------------------
 * File Name   : CCCameraAppUI.js
 * Description : AppUI Scene camera functionality.
 *
 * Created     : 20/10/11
 * Author(s)   : Ashraf Samy Hegab
 *-----------------------------------------------------------
 */

function CCCameraAppUI()
{
    this.construct();
	
	this.LookAtTarget = vec3.create();
	this.offsetTarget = vec3.create();
	
	this.lookAtInterpolator = new CCInterpolatorCurveV3( CCInterpolatorCurveX2 ).setup( this.lookAt );
	var offsetInterpolator = this.offsetInterpolator = new CCInterpolatorCurveV3( CCInterpolatorCurve ).setup( this.offset );
	offsetInterpolator.setDuration( 0.3 );
}
CopyPrototype( CCCameraAppUI, CCCameraBase, "CCCameraBase" );


CCCameraAppUI.prototype.interpolateLookAt = function(delta, speed)
{
    var updating = false;
	
	var offsetInterpolator = this.offsetInterpolator;
	var offsetTarget = this.offsetTarget;	
    var targetOffset = this.targetOffset;
	var offset = this.offset;
	if( !vec3.equals( offsetTarget, targetOffset ) )
	{
		vec3.set( targetOffset, offsetTarget );
		offsetInterpolator.setup( offset, offsetTarget );
	}
	
	if( offsetInterpolator.update( delta ) )
	{
		updating = true;
	}
	
	var lookAtInterpolator = this.lookAtInterpolator;
	var LookAtTarget = this.LookAtTarget;	
    var targetLookAt = this.targetLookAt;
	var lookAt = this.lookAt;
	if( !vec3.equals( LookAtTarget, targetLookAt ) )
	{
		vec3.set( targetLookAt, LookAtTarget );
		lookAtInterpolator.setup( lookAt, LookAtTarget );
	}
	
	if( lookAtInterpolator.update( delta * speed ) )
	{
		updating = true;
	}
    
    if( updating )
    {
        var position = this.position;
        vec3.set( lookAt, position );
        position[2] += offset[2];
    }
    
    return false;
}

File: 3.cameraandmatrices/mobile/Engine/Source/Tools/CCCameraAppUI.cpp
/*-----------------------------------------------------------
 * 2c - Cross Platform 3D Application Framework
 *-----------------------------------------------------------
 * Copyright © 2010 – 2011 France Telecom
 * This software is distributed under the Apache 2.0 license,
 * see the "license.txt" file for more details.
 *-----------------------------------------------------------
 * File Name   : CCCameraAppUI.cpp
 *-----------------------------------------------------------
 */

#include "CCDefines.h"


CCCameraAppUI::CCCameraAppUI()
{
    offsetInterpolator.setDuration( 0.3f );
}


void CCCameraAppUI::update()
{
    super::update();
}


const bool CCCameraAppUI::interpolateCamera(const float delta, const float speed)
{
	if( updating )
	{   
        updating = false;
        
        if( currentOffsetTarget.target != targetOffset )
        {
            currentOffsetTarget.target = targetOffset;
            offsetInterpolator.setup( &currentOffsetTarget.current, currentOffsetTarget.target );
        }
        
        if( currentLookAtTarget.target != targetLookAt )
        {
            currentLookAtTarget.target = targetLookAt;
            lookAtInterpolator.setup( &currentLookAtTarget.current, currentLookAtTarget.target );
        }
        
        if( offsetInterpolator.update( delta ) )
        {
            refreshCameraSize();
            updating = true;
        }
        
        if( lookAtInterpolator.update( delta * speed ) )
        {
            updating = true;
            updatedPosition = true;
        }
        
        // Now update our camera
        {
            lookAt = currentLookAtTarget.current;
            updateOffset( currentOffsetTarget.current );
        }
        return true;
	}

	return false;
}


void CCCameraAppUI::setOffset(const CCVector3 &offsetTarget)
{
    currentOffsetTarget.current = currentOffsetTarget.target = offsetTarget;
    updateOffset( offsetTarget );
}


void CCCameraAppUI::setCameraWidth(const float inWidth)
{
    cameraWidth = inWidth;
    targetOffset.z = cameraWidth;
    if( gEngine->renderer->isPortrait() )
    {
        targetOffset.z /= getFrustumSize().width;
        cameraHeight = targetOffset.z * getFrustumSize().height;
    }
    else
    {
        targetOffset.z /= getFrustumSize().height;
        cameraHeight = targetOffset.z * getFrustumSize().width;
    }
    
    cameraHWidth = cameraWidth * 0.5f;
    cameraHHeight = cameraHeight * 0.5f;
}


void CCCameraAppUI::setCameraHeight(const float inHeight)
{
    cameraHeight = inHeight;
    targetOffset.z = cameraHeight;
    if( gEngine->renderer->isPortrait() )
    {
        targetOffset.z /= getFrustumSize().height;
    }
    else
    {
        targetOffset.z /= getFrustumSize().width;
    }
    
    cameraWidth = targetOffset.z * getFrustumSize().height;
    cameraHWidth = cameraWidth * 0.5f;
    cameraHHeight = cameraHeight * 0.5f;
}


void CCCameraAppUI::refreshCameraSize()
{
    CCVector3 &currentOffset = currentOffsetTarget.current;
    if( gEngine->renderer->isPortrait() )
    {
        cameraWidth = currentOffset.z * getFrustumSize().width;
        cameraHeight = currentOffset.z * getFrustumSize().height;
    }
    else
    {
        cameraWidth = currentOffset.z * getFrustumSize().height;
        cameraHeight = currentOffset.z * getFrustumSize().width;
    }
    
    cameraHWidth = cameraWidth * 0.5f;
    cameraHHeight = cameraHeight * 0.5f;
}